# Study of borrower reliability


## Open the table and explore general information about the data


**Task 1. Import the pandas library. Read data from a csv file into a dataframe and save it into the `data` variable. The path to the file:** `/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Task 2. Display the first 20 lines of the `data` dataframe on the screen.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Task 3. Display basic information about the dataframe using the method `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Data preprocessing

### Removing gaps

**Task 4. Print the number of missing values ​​for each column. Use a combination of two methods.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Task 5. Two columns have missing values. One of them is `days_employed`. You will handle gaps in this column in the next step. Another column with missing values, `total_income`, stores income data. The amount of income is most influenced by the type of employment, so you need to fill in the gaps in this column with the median value for each type from the `income_type` column. For example, for a person with an employment type of `employee`, the gap in the `total_income` column should be filled with the median income among all records with the same type.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Handling anomalous values

**Task 6. Data may contain artifacts (anomalies) - values ​​that do not reflect reality and appeared due to some error. such an artifact would be a negative number of days of employment in the `days_employed` column. This is normal for real data. Process the values ​​in this column: replace all negative values ​​with positive ones using the method `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Task 7. For each type of employment, print the median value of length of service days_employed in days.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

Two types (unemployed and pensioners) will have abnormally large values. These values ​​are difficult to correct, so leave them as they are. Moreover, you will not need this column for research.

**Task 8. List the unique values ​​of a column `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Task 9. There are two anomalous values ​​in the `children` column. Remove rows containing such anomalous values ​​from the dataframe `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Task 10. List the unique values ​​of the `children` column again to make sure that the artifacts have been removed.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Removing gaps (continued)

**Task 11. Fill in the blanks in the `days_employed` column with the median values ​​for each type of employment `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Task 12. Make sure all blanks are filled in. Test yourself and again print the number of missing values ​​for each column using two methods.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Changing Data Types

**Task 13. Replace the real data type in the `total_income` column with an integer using the method `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Handling duplicates

**Task 14: Handle implicit duplicates in the `education` column. This column has the same values, but written differently, using uppercase and lowercase letters. Convert them to lower case. Check the remaining columns.**

In [14]:
data['education'] = data['education'].str.lower()

**Task 15. Display the number of duplicate rows in the data. If such lines are present, remove them.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Data categorization

**Task 16. Based on the ranges indicated below, create a column `total_income_category` with categories in the `data` dataframe:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**For example, a borrower with an income of 25,000 should be assigned category `'E'`, and a client receiving 235,000 should be assigned category `'B'`. Use your own function called `categorize_income()` and method `apply()`.**

In [26]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E (< 30 000)'
        elif 30001 <= income <= 50000:
            return 'D (30 001 - 50 000)'
        elif 50001 <= income <= 200000:
            return 'C (50 001 - 200 000)'
        elif 200001 <= income <= 1000000:
            return 'B (200 001 - 1 000 000)'
        elif income >= 1000001:
            return 'A (more 1 000 000)'
    except:
        pass

In [27]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Task 17. Display a list of unique purposes for taking out a loan from the column `purpose`.**

In [28]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Task 18. Create a function that, based on data from the `purpose` column, will generate a new `purpose_category` column, which will include the following categories:**

- ``operations with a car'`,
- ``real estate transactions'`,
- ``carrying out a wedding'`,
- ``getting an education'`.

**For example, if the `purpose` column contains the substring ``for the purchase of a car'`, then the `purpose_category` column should contain the string ``operations with a car'`.**

**Use your own function named `categorize_purpose()` and method `apply()`. Examine the data in the `purpose` column and determine which substrings will help you correctly identify the category**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'operations with a car'
        elif 'жил' in row or 'недвиж' in row:
            return 'real estate transactions'
        elif 'свад' in row:
            return 'carrying out a wedding'
        elif 'образов' in row:
            return 'getting an education'
    except:
        return 'mo category'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

#### Is there a relationship between the number of children and repayment of the loan on time?

In [22]:
# We will group the data using the pivot table method by the number of children, in the columns we will calculate the number of loans issued to borrowers,
# with the corresponding number of children and, immediately, the share of overdue loans.
# We will format the output of the share of overdue loans: round to 4 decimal places and multiply by 100.

result_3_1 = pd.pivot_table(data,index=['children'],values =['debt'], aggfunc=['count','mean'])
result_3_1[( 'mean', 'debt')]=result_3_1[( 'mean', 'debt')].round(4)*100

display(result_3_1)

count  mean
           debt  debt
children             
0         14091  7.54
1          4808  9.23
2          2052  9.45
3           330  8.18
4            41  9.76
5             9  0.00

**Conclusion:**

From the data provided, it is clear that if there are 5 children, there are no overdue loan payments; at the same time, there are only 9 such loans, which means that their influence on the overall conclusion can be neglected. The share of overdue loans from the total number of loans for borrowers with 1, 2, 4 children is approximately equal and ranges from 9.2% to 9.7%. The share of overdue loans from the total number of loans for borrowers with 3 children is comparatively lower and amounts to 8.1%. The share of overdue loans from the total number of loans for borrowers without children is the lowest of all categories of borrowers and amounts to 7.5%.

**General conclusion:** the very fact that borrowers have children increases the likelihood of loan delinquency, while the number of children has less influence on the change in this probability.

#### Is there a relationship between marital status and repayment of the loan on time?

In [23]:
# We will group the data using the pivot table method by marital status, and in the columns we will calculate the number of loans issued to borrowers,
# according to the relevant marital status and, immediately, the share of overdue loans.
# We will format the output of the share of overdue loans: round to 4 decimal places and multiply by 100.

result_3_2 = pd.pivot_table(data,index=['family_status'],values =['debt'], aggfunc=['count','mean'])
result_3_2[( 'mean', 'debt')]=result_3_2[( 'mean', 'debt')].round(4)*100

display(result_3_2)

count  mean
                        debt  debt
family_status                     
Не женат / не замужем   2796  9.76
в разводе               1189  7.06
вдовец / вдова           951  6.62
гражданский брак        4134  9.31
женат / замужем        12261  7.56

**Conclusion:**

From the presented data it is clear that borrowers with formalized family relationships (married/divorced, widower) have a lower percentage of overdue loans; this percentage is in the range of 6.6% - 7.5%. At the same time, there is a tendency towards a decrease in this percentage for borrowers with ended family relationships: 7.0% for divorced people, versus 7.5% for married ones. For borrowers who are not in a formal relationship (civil marriage, single), the share of overdue loans is significantly higher - 9.3% - 9.7% compared to borrowers with a formal marital status.

**General conclusion:** if the borrower has a formalized family relationship, it will increase the likelihood of timely repayment of the loan. At the same time, the greatest discipline in terms of returns on time is observed in the “widower/widow” category. And for borrowers with unregistered family relationships (civil marriage), the probability of delay is comparable to unmarried borrowers.

#### Is there a relationship between income level and loan repayment on time?

In [29]:
# We will group the data using the pivot table method by income level category, and in the columns we will calculate the number of loans issued
# borrowers, according to their marital status and, immediately, the share of overdue loans.
# We will format the output of the share of overdue loans: round to 4 decimal places and multiply by100.

result_3_3 = pd.pivot_table(data,index=['total_income_category'],values =['debt'], aggfunc=['count','mean'])
result_3_3[( 'mean', 'debt')]=result_3_3[( 'mean', 'debt')].round(4)*100

print(result_3_3)

                         count  mean
                          debt  debt
total_income_category               
A (more 1 000 000)          25  8.00
B (200 001 - 1 000 000)   5014  7.06
C (50 001 - 200 000)     15921  8.50
D (30 001 - 50 000)        349  6.02
E (< 30 000)                22  9.09


**Conclusion:**

From the presented data it is clear that the most disciplined are borrowers with income category “D” - from 30,001 to 50,000. For them, the share of overdue loans is a minimum of 6.0%. The least disciplined are borrowers with income category "E" - up to 30,000. For them, the share of overdue loans is 9.0%. In other categories, the share of overdue loans is 7.0 - 8.4%

**General conclusion:** The probability of loan delinquency is minimal for borrowers with an income level from 30,001 to 50,000 and maximum for borrowers with an income level up to 30,000. For other groups of borrowers, there is a weak dependence of the probability of loan delinquency on income level.

#### How do different purposes of a loan affect its repayment on time?

In [25]:
# We will group the data using the pivot table method by category of loan purpose, and in the columns we will calculate the number of loans issued
# to borrowers, according to the corresponding category of loan purpose and, immediately, the share of overdue loans.
# We will format the output of the share of overdue loans: round to 4 decimal places and multiply by 100.

result_3_4 = pd.pivot_table(data,index=['purpose_category'],values =['debt'], aggfunc=['count','mean'])
result_3_4[( 'mean', 'debt')]=result_3_4[( 'mean', 'debt')].round(4)*100

display(result_3_4)

count  mean
                           debt  debt
purpose_category                     
carrying out a wedding     2313  7.91
getting an education       3988  9.25
operations with a car      4279  9.35
real estate transactions  10751  7.26

**Conclusion:**

From the presented data it is clear that the least risky are loans associated with the purchase of real estate. In this category, on average, the share of overdue loans is 7.2%. The most risky are loans for the purchase of a car. Here the share of overdue loans is more than 9.3%. Education loans are only slightly inferior in risk to car loans. The overdue rate for them is 9.2%

**General conclusion**: loans for the purchase of real estate, on average, are the most reliable, and for car loans and education loans, in comparison, the risk of delinquency increases by approximately 1.3 times.

#### Give possible reasons for the appearance of gaps in the source data.

*Answer:*

As can be seen from paragraph 3. Data gaps occurred only in the “Work Experience” and “Amount of Income” columns.
Upon further examination of the data (clause 1.), you can note that the omissions are typical for “Income Type” = “pensioner”, i.e., it can be assumed that the presence of omissions of data is due to the inaccuracy of the loan application form, which allows When filling out, do not enter data in the specified columns; when choosing the type of income = “pensioner”.

**In general**, gaps in data may be the result of errors in the operation of data recording devices, errors in data transfer, export/import, errors in algorithms for generating data downloads, errors in data entry, or intentional omission of certain data when entering.
    
    
Additionally, omissions **in the studied dataset** may be the result of either erroneously not filling out this data (missed, not noticed), or deliberately ignoring it, either due to the fact that the potential lender did not have a job at the time of applying for a loan and, therefore, had no income i.e. or was temporarily unemployed or, in principle, had never worked before, in this case the lack of data on work experience and earnings is fully consistent with the real state of affairs.

#### Explain why filling the blanks with the median is the best solution for quantitative variables.

*Answer:*

It is best to fill in the gaps in quantitative variables with median values, because Such filling of gaps, firstly, allows you to save a dataframe row for further analysis in other columns and secondly, it has the least impact (introduces the least error) into the statistical characteristics of the column under study, incl. and because for data that has a normal distribution (or close to normal), with a relatively large amount of data, the median will be equal to the mean value.

The median is a value that splits the data, **sorted in order** of increasing values, into two equal parts. That is, the median shows the central value in the sample if the number of observations is odd and the arithmetic mean of two values if the number of observations in the sample is even. That. To determine the median, you must first sort the list of data and then take the middle (located in the middle of the list) data element. While the average is the result of dividing the sum of all data by their number.
    
**The main difference** between the median and the average is its resistance to “outliers” - uncharacteristic of the sample being studied, values that are too large or too small. Such "outliers" can have a significant impact on the average even if the number of "outliers" themselves is insignificant. And, on the contrary, they do not affect the median.

### General conclusion.

The purpose of this project was to study the reliability of borrowers, depending on various factors, incl. marital status, number of children, income level. The data volume was 21,525 rows. After preprocessing and excluding rows with anomalous values, the data volume totaled 21,331 rows.

Based on the results of the study, conclusions were drawn about the dependence of the probability of loan delays on various characteristics of the borrower.

Based on these findings, it is possible to draw up a “portrait of an **ideal** borrower”, in respect of whom one can expect the least probability of delay. This is a borrower who is/was in a registered marriage, without children, with an income level of 30,001 to 50,000, who wants to get a loan to purchase real estate.

And, conversely, the **least reliable** borrower will be an unmarried borrower, incl. civilian, with 4 children, with an income of up to 30,000, wanting to get a car loan.